### DISCUSSION TITLE
Predefined Rows as Header or Footer on Filament Table

### DISCUSSION THREAD

**Discussion by kombiahrk** [filamentphp/filament#9421] [Help]
While listing the records need to show some custom rows is it possible in Filament

**Comment by sinnbeck** [comment]
https://filamentphp.com/docs/3.x/tables/summaries

But I think it only goes below the table, not above. What do you need above?

**Reply by kombiahrk** [reply]
i need to display opening balance, with summary some how i achieved this but while grouping it showing incorrect opening balance.

**Reply by sinnbeck** [reply]
Sadly it's hard to debug as you haven't posted any code

**Reply by kombiahrk** [reply]
First of thanks for quick reply

class OrderItemRelationManager extends RelationManager
{
    protected static string $relationship = 'items';

    protected static ?string $title = 'Order Details';

    public function table(Table $table): Table
    {
        return $table
            ->recordTitleAttribute('order_id')
            ->columns([
                Tables\Columns\TextColumn::make('order.order_date')->label('Order Date')->date('d/m/Y'),
                Tables\Columns\TextColumn::make('order.order_number')->label('Order Number'),
                Tables\Columns\TextColumn::make('order.party.name')->label('Party Name'),
                Tables\Columns\TextColumn::make('quantity')
                    ->summarize(Summarizer::make()
                        ->label('Opening Balance')
                        ->using(function (Builder $query): string {

                            $bindings = $query->getBindings(); // Get all the bindings
                            $firstBinding = reset($bindings);
                            //dd($query);
                            $productId = $firstBinding; // Replace with the actual product ID
                            $product = Product::find($productId);

                            $currentQuantity = $product->quantity;

                            $query1 = clone $query;
                            $query2 = clone $query;
                            $query3 = clone $query;

                            if (!$product->enable_alt_units) {
                                $openingBalanceIn = $query->where('order_type', 'pur')->sum('quantity');
                                $openingBalanceOut = $query1->where('order_type', 'sale')->sum('quantity');

                                $openingBalance = $currentQuantity - ($openingBalanceIn - $openingBalanceOut);
                            } else {
                                // Calculate the opening balance for the alternate unit

                                $unit = ($product->unit_id);
                                $altUnit = ($product->alt_unit_id);

                                $openingBalanceInAlt = $query->where('order_type', 'pur')->where('unit_id', $altUnit)->sum('quantity');
                                $openingBalanceOutAlt = $query1->where('order_type', 'sale')->where('unit_id', $altUnit)->sum('quantity');

                                $openingBalanceInBase = $query2->where('order_type', 'pur')->where('unit_id', $unit)->sum('quantity');
                                $openingBalanceOutBase = $query3->where('order_type', 'sale')->where('unit_id', $unit)->sum('quantity');
                                // Convert the opening balance for the alternate unit to the base unit
                                $openingBalanceInBase += ($openingBalanceInAlt / $product->alt_unit_value);
                                $openingBalanceOutBase += ($openingBalanceOutAlt / $product->alt_unit_value);

                                $openingBalance = $currentQuantity - ($openingBalanceInBase - $openingBalanceOutBase);
                                    echo '<pre>';
                                    print_r($openingBalance.' '.$currentQuantity.' '.$openingBalanceInBase.' '.$openingBalanceOutBase);
                                    echo '</pre>';
                            }
                            if ($product->enable_alt_units) {
                                $unit = Unit::find($product->unit_id);
                                $altUnit = Unit::find($product->alt_unit_id);
                                $integerPart = floor($openingBalance);
                                $decimalPart = ($openingBalance - $integerPart) * $product->alt_unit_value;
                                return $integerPart . ' ' . $unit->name . ' ' . $decimalPart . ' ' . $altUnit->name;
                            } else {
                                $unit = Unit::find($product->unit_id);
                                return floor($openingBalance) . ' ' . $unit->name;
                            }
                        }))
                    ->state(function (Model $record): string {
                        //dd($record);
                        if ($record->enable_alt_units) {
                            $unit = Unit::find($record->unit_id);
                            $altUnit = Unit::find($record->alt_unit_id);
                            $integerPart = floor($record->quantity);
                            $decimalPart = ($record->quantity - $integerPart) * $record->alt_unit_value;
                            return $integerPart . ' ' . $unit->short_name . ' ' . $decimalPart . ' ' . $altUnit->short_name;
                        } else {
                            $unit = Unit::find($record->unit_id);
                            return floor($record->quantity) . ' ' . $unit->short_name;
                        }
                    }),
            ])
            ->groups([
                Group::make('order.order_date')->label('Order Date')->date('d/m/Y'),
                Group::make('order.party.name')->label('Party Name'),
            ])
            ->defaultSort('order_id', 'desc')
            ->recordAction(null)
            ->filters([
                //
            ])
            ->headerActions([
                // Tables\Actions\CreateAction::make(),
            ])
            ->actions([
                //Tables\Actions\EditAction::make(),
                //Tables\Actions\DeleteAction::make(),
            ])
            ->bulkActions([
                Tables\Actions\BulkActionGroup::make([
                    // Tables\Actions\DeleteBulkAction::make(),
                ]),
            ]);
    }
}

and here is the code
